cmd/cloud_upload.go

```diff
@@ -0,0 +1,189 @@
+package cmd
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+	"github.com/fatih/color"
+	"github.com/spf13/cobra"
+	"github.com/spf13/pflag"
+	"path/filepath"
+
+	"go.k6.io/k6/cloudapi"
+	"go.k6.io/k6/cmd/state"
+	"go.k6.io/k6/lib"
+	"go.k6.io/k6/lib/consts"
+	"go.k6.io/k6/ui/pb"
+)
+
+// cmdCloudUpload handles the `k6 cloud upload` sub-command
+type cmdCloudUpload struct {
+	gs *state.GlobalState
+}
+
+func getCmdCloudUpload(gs *state.GlobalState) *cobra.Command {
+	c := &cmdCloudUpload{
+		gs: gs,
+	}
+
+	exampleText := getExampleText(gs, `
+  # Upload the test to the k6 Cloud without actually starting a test run
+  $ {{.}} cloud upload script.js
+
+  # Upload the test to the k6 Cloud with a specific token
+  $ {{.}} cloud upload --token <YOUR_API_TOKEN> script.js`[1:])
+
+	cloudRunCmd := &cobra.Command{
+		Use:   cloudUploadCommandName,
+		Short: "Upload the test to the Grafana k6 Cloud",
+		Long: `Upload the test to the Grafana k6 Cloud.
+
+This will upload the test to the Grafana k6 Cloud service. Using this command requires to be authenticated
+against the Grafana k6 Cloud. Use the "k6 cloud login" command to authenticate.`,
+		Example: exampleText,
+		Args:    exactArgsWithMsg(1, "arg should either be \"-\", if reading script from stdin, or a path to a script file"),
+		RunE:    c.run,
+	}
+
+	cloudRunCmd.Flags().SortFlags = false
+	cloudRunCmd.Flags().AddFlagSet(c.flagSet())
+
+	return cloudRunCmd
+}
+
+func (c *cmdCloudUpload) run(cmd *cobra.Command, args []string) error {
+	printBanner(c.gs)
+
+	progressBar := pb.New(
+		pb.WithConstLeft("Init"),
+		pb.WithConstProgress(0, "Loading test script..."),
+	)
+	printBar(c.gs, progressBar)
+
+	test, err := loadAndConfigureLocalTest(c.gs, cmd, args, getPartialConfig)
+	if err != nil {
+		return err
+	}
+
+	// It's important to NOT set the derived options back to the runner
+	// here, only the consolidated ones. Otherwise, if the script used
+	// an execution shortcut option (e.g. `iterations` or `duration`),
+	// we will have multiple conflicting execution options since the
+	// derivation will set `scenarios` as well.
+	testRunState, err := test.buildTestRunState(test.consolidatedConfig.Options)
+	if err != nil {
+		return err
+	}
+
+	// TODO: validate for usage of execution segment
+	// TODO: validate for externally controlled executor (i.e. executors that aren't distributable)
+	// TODO: move those validations to a separate function and reuse validateConfig()?
```

Comment:
Most of this code, as well as some of the code below, can be shared with `k6 cloud run`. Perhaps, we could try to extract some common functions, also to make these very long methods a bit more readable?